src/main/java/org/elasticsearch/common/property/PromptPlaceholderResolver.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.property;
+
+import com.google.common.collect.UnmodifiableIterator;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.cli.Terminal;
+import org.elasticsearch.common.settings.ImmutableSettings;
+import org.elasticsearch.common.settings.Settings;
+
+import java.util.Map;
+
+/**
+ * Utility class for replacing prompt placeholders <code>__prompt__</code> in settings by prompting a user for the actual
+ * setting value.
+ */
+public final class PromptPlaceholderResolver {
+
+    public static final String PROMPT_VALUE = "__prompt__";
```

Comment:
I think we should keep it generic. I think adding `__prompt:secret__` and `__prompt:text__` is the right thing to do. However, I don't see the need for other types since we store settings internally as a string or maybe I am missing something here?